### PR TITLE
Fix pulumi's resources urns completion

### PR DESCRIPTION
**Pull request by brunoroque06**
Use `pulumi stack export` instead of `pulumi stack --show-urns` to get the resources' urns. The former outputs a `json`, whereas the latter outputs a table (maybe more user friendly?).

I think there was an additional `carapace.ActionCallback(...)` call as well. Or was it intentional? The current implementation does not work for me; it completes with the single entry `urn`.

### CODE REVIEW

**Comment by rsteube** [comment]
> I think there was an additional carapace.ActionCallback(...) call as well. Or was it intentional? The current implementation does not work for me; it completes with the single entry urn.

Not sure, a while ago that i've written it. Either a mistake or might be related to the caching - will have a look at it.

> Use pulumi stack export instead of pulumi stack --show-urns to get the resources' urns. The former outputs a json, whereas the latter outputs a table (maybe more user friendly?).

Not sure why I've used the other, `json` is definitely better.

I only tested it against a local docker (:thinking:) so it's likely there are some mistakes that become apparent in productive use.

**Comment by brunoroque06** [comment]
Yes, this one is a bit annoying to test because, as far as I know, one needs to:

- create a pulumi project;
- create a stack;
- deploy something somewhere.

So that `pulumi stack export` returns whatever pulumi has stored regarding our project/stack.

But this implementation works for me. I did not run into any issue :x don't know if I missed something. I tested the commands `pulumi up -t <urn>`, and `pulumi state delete <urn>`.

**Comment by rsteube** [comment]
Hard to see right now even for me (who wrote it). I think it was or at least was supposed to be (might be superflous) for the caching.
Which I think we might remove as well. It's just 5 seconds and I think that was for the `ToMultiPartsA("::")` which completes the segments separately and thus invokes the action on each.
As far as I remember the urn's were quite long? Think there was a reason why I added that but if the completion is fine on your side we should just leave it out.

**Comment by rsteube** [comment]
> But this implementation works for me. I did not run into any issue :x don't know if I missed something. I tested the commands pulumi up -t <urn>, and pulumi state delete <urn>.

When you say it is working fine that is good enough for me. I personally lack the experience with pulumi to test it in depth.
Any mistake can be fixed later when it becomes obvious.

File: completers/pulumi_completer/cmd/action/urn.go [hunk 1]
```diff
@@ -1,60 +1,53 @@
...
+			var stack Stack
+			if err := json.Unmarshal(output, &stack); err != nil {
+				return carapace.ActionMessage(err.Error())
+			}
+			urns := make([]string, len(stack.Deployment.Resources))
+			for i, r := range stack.Deployment.Resources {
+				urns[i] = r.Urn
+			}
+			return carapace.ActionValues(urns...)
+		}).Cache(5*time.Second,
```
Comment (rsteube) [inline completers/pulumi_completer/cmd/action/urn.go:36]: I think we can remove the caching. Unless you experience slowness somewhere, but i don't see repetitive calls to it at the moment where the short cache would help :thinking:

**Review by rsteube — CHANGES_REQUESTED** [review]

**Comment by brunoroque06** [comment]
> As far as I remember the urn's were quite long? Think there was a reason why I added that but if the completion is fine on your side we should just leave it out.

Yes, they are long. This is an example of an `urn` on Azure:

`urn:pulumi:<stack>::<project>::azure-native:storage:StorageAccount::<account>`

I don't think the `ToMultiPartsA("::")` makes a lot of sense here, because all of the `urn`s, for a given project/stack, will have the same preffix: `urn:pulumi:<stack>::<project>::azure-native:*`. Also notice that it has both `::` and `:` as separators. So we just want to complete everything until `azure-native` in one `<tab>` press, and then go from there.

The cache works, and I think it is good and makes sense here. It takes some time to execute `pulumi stack export`, as it is a call to their backend.

File: completers/pulumi_completer/cmd/action/urn.go (hunk 1 above)
Comment (brunoroque06) [reply completers/pulumi_completer/cmd/action/urn.go:36]: Just tested the cache again, and I think it is a good thing to have. In my case it usually takes 2 `<tab>`s to complete a `urn` (in `elvish). Each `pulumi stack export` takes ~ 2 seconds. I would say it is even safe to increase the cache timeout, as these are typically slow changing resources.

**Comment by rsteube** [comment]
Ah right, so that's the shell completion up to there and thus the subsequent call to carapace. Makes sense then.

File: completers/pulumi_completer/cmd/action/urn.go (hunk 1 above)
Comment (rsteube) [reply completers/pulumi_completer/cmd/action/urn.go]: Yeah, increase it to what you think makes sense (don't forget about newly created resources).

**Review by rsteube — APPROVED** [review]

File: completers/pulumi_completer/cmd/action/urn.go (hunk 1 above)
Comment (brunoroque06) [reply completers/pulumi_completer/cmd/action/urn.go]: Apologies, missed this comment. Thank you!

**Pull request merged**